Moved creating SiPixelTemplateStore from DB to an ESProducer [13_2]

Dr15Jones opened the pull request:
#### PR description:

All DB construction was switched to using EventSetup. Cases using reading from files were kept local to module.
The new ESProducer was added as a Task to all needed Sequences.

#### PR validation:

Code compiles. 

backport of #42514 and #42597

Dr15Jones committed:
Moved creating SiPixelTemplateStore from DB to an ESProducer

All DB construction was switched to using EventSetup. Cases using
reading from files were kept local to module.
The new ESProducer was added as a Task to all needed Sequences.

Dr15Jones committed:
Do not use a Task with SiPixelTemplateStoreESProducer

The use of a Task at this time causes to much disruption. Moving
ES modules to Tasks should wait till a dedicated campaign.

cmsbuild commented:
A new Pull Request was created by  @Dr15Jones (Chris Jones) for CMSSW_13_2_X.

It involves the following packages:

- Alignment/OfflineValidation (**alca**)
- CalibTracker/SiPixelLorentzAngle (**alca**)
- Calibration/TkAlCaRecoProducers (**alca**)
- CondFormats/SiPixelTransient (**db, reconstruction**)
- DQM/SiPixelPhase1Config (**dqm**)
- DQM/SiPixelPhase1Track (**dqm**)
- DQMOffline/Trigger (**dqm**)
- FastSimulation/Configuration (**fastsim**)
- FastSimulation/TrackingRecHitProducer (**fastsim**)
- HLTrigger/Configuration (**hlt**)
- RecoHI/HiTracking (**reconstruction**)
- RecoLocalTracker/SiPixelRecHits (**reconstruction**)
- RecoTracker/FinalTrackSelectors (**reconstruction**)
- RecoTracker/TrackProducer (**reconstruction**)
- RecoTracker/TransientTrackingRecHit (**reconstruction**)


@perrotta, @Martin-Grunewald, @ssekmen, @tvami, @nothingface0, @civanch, @emanueleusai, @consuegs, @mdhildreth, @mandrenguyen, @cmsbuild, @missirol, @saumyaphor4252, @clacaputo, @syuvivida, @sbein, @pmandrik, @tjavaid, @micsucmed, @francescobrivio, @rvenditti can you please review it and eventually sign? Thanks.
@VourMa, @felicepantaleo, @Martin-Grunewald, @Fedespring, @tlampen, @threus, @mmusich, @seemasharmafnal, @silviodonato, @JanFSchulte, @jhgoh, @missirol, @yenjie, @cericeci, @ferencek, @trocino, @yetkinyilmaz, @fioriNTU, @sroychow, @GiacomoSguazzoni, @rovere, @VinInn, @mroguljic, @tocheng, @jandrea, @idebruyn, @ebrondol, @mtosi, @dgulhan, @arossi83, @HuguesBrun, @adewit, @mandrenguyen, @jazzitup, @dkotlins, @kurtejung, @gpetruc, @matt-komm, @rociovilar, @tvami this is something you requested to watch as well.
@perrotta, @dpiparo, @antoniovilela, @rappoccio you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

mmusich commented:
this branch has conflicts to be resolved.

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-67adef/34452/summary.html
**COMMIT**: 1e1d2a0cb6a229ceb11676abef82b60ad7018db1
**CMSSW**: CMSSW_13_2_X_2023-08-23-1100/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/42644/34452/install.sh` to create a dev area with all the needed externals and cmssw changes.



The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
 - @CMSTrackingPOG cms-sw/cmssw#42593

You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-67adef/34452/git-recent-commits.json
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-67adef/34452/git-merge-result


## Comparison Summary



**Summary**:
 - You potentially added 4 lines to the logs
 - Reco comparison results: 8 differences found in the comparisons
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3196338
 - DQMHistoTests: Total failures: 3
 - DQMHistoTests: Total nulls: 4
 - DQMHistoTests: Total successes: 3196309
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 55.254 KiB( 47 files compared)
 - DQMHistoSizes: changed ( 12434.0,... ): 27.627 KiB       HLT/Tracking
 - Checked 207 log files, 159 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_2_X_2023-08-23-1100+67adef/58575/triggerResults): no differences found

missirol commented:
+hlt

(This update will require the 'parsing' of the next CMSSW_13_2_X release into ConfDB, in order to add the new ESProducer to the HLT menus in the db, I think.)

missirol commented:
backport of https://github.com/cms-sw/cmssw/pull/42514
backport of https://github.com/cms-sw/cmssw/pull/42597

perrotta commented on Calibration/TkAlCaRecoProducers/python/ALCARECOPromptCalibProdSiPixelLorentzAngleMCS_cff.py:
```diff
@@ -26,6 +26,7 @@
 from RecoTracker.Configuration.RecoTrackerP5_cff import *
 from RecoTracker.TrackProducer.TrackRefitter_cfi import *
 from TrackingTools.TransientTrack.TransientTrackBuilder_cfi import *
+from RecoLocalTracker.SiPixelRecHits.SiPixelTemplateStoreESProducer_cfi import *
```
I should have had commented already in #42597...

But as of now these `include`'s in the modified Calibration/TkAlCaRecoProducers/python configs are not needed any more. Should they get removed, or do you plan to let them there as a placeholder for future reinsertion of Tasks in them?

Dr15Jones replied:
I don't have any (short term) plans on reinserting Tasks.

perrotta replied:
> I don't have any (short term) plans on reinserting Tasks.

Thank you @Dr15Jones 
Should we then clean up these backports from those unnecessary includes, together with the master (with an additional PR)?

mmusich replied:
yes - please.

perrotta commented on HLTrigger/Configuration/python/customizeHLTforCMSSW.py:
```diff
@@ -215,12 +215,20 @@ def customiseForOffline(process):
     return process
 
 
+def customizeHLTfor42514(process):
```
@missirol @Martin-Grunewald I don't remember: should we put here the number of this PR, or the one in the master (as  implemented here)?

Martin-Grunewald replied:
In principle the one for the specific release...

missirol replied:
> In principle the one for the specific release...

True, but I would suggest to keep the number of the original PR. The reason is that keeping the files identical (as much as possible) across cycles helps when cherry-picking, avoiding silly conflicts.

mmusich replied:
@missirol @Martin-Grunewald I assume from the discussion above that we are fine with keeping the old release number. Please shout if not the case.

perrotta reviewed (commented):
The three backports in 13_2_X, 13_1_X and 13_0_X are identical, and all verbatim backports of the combination of #42514 and #42597. Therefore the same comments made here are also valid for #42647 and #42648

cmsbuild commented:
Pull request #42644 was updated. @perrotta, @Martin-Grunewald, @ssekmen, @tvami, @nothingface0, @civanch, @emanueleusai, @consuegs, @mdhildreth, @mandrenguyen, @cmsbuild, @missirol, @saumyaphor4252, @clacaputo, @syuvivida, @sbein, @pmandrik, @tjavaid, @micsucmed, @francescobrivio, @rvenditti can you please check and sign again.

mmusich commented on Calibration/TkAlCaRecoProducers/python/ALCARECOPromptCalibProdSiPixelLorentzAngle_cff.py:
```diff
@@ -62,9 +62,10 @@
                                                    )
 
 # The actual sequence
+from RecoLocalTracker.SiPixelRecHits.SiPixelTemplateStoreESProducer_cfi import *
```
is it still needed here?

mmusich commented on Calibration/TkAlCaRecoProducers/python/ALCARECOPromptCalibProdSiStripGainsAAG_cff.py:
```diff
@@ -63,7 +63,8 @@
 # refit and BS can be dropped if done together with RECO.
-# track filter can be moved in acalreco if no otehr users
+# track filter can be moved in acalreco if no other users
+from RecoLocalTracker.SiPixelRecHits.SiPixelTemplateStoreESProducer_cfi import SiPixelTemplateStoreESProducer
```
here too?

mmusich commented on Calibration/TkAlCaRecoProducers/python/ALCARECOSiPixelCalSingleMuonTight_cff.py:
```diff
@@ -76,6 +76,7 @@
 ##################################################################
 # Sequence: track refit + close-by-pixel producer
 ##################################################################
+from RecoLocalTracker.SiPixelRecHits.SiPixelTemplateStoreESProducer_cfi import SiPixelTemplateStoreESProducer
```
and here?

perrotta commented:
> Removing all the unnecessary `imports` without first testing them in the master is beyond what I'm willing to do. A domain expert should take over if a different backport is wanted.

Indeed, I'd remove them for master first, and then backport here.
I can do the master PR tomorrow (this evening is already too late for me)

mmusich committed:
remove a bunch of useless includes of SiPixelTemplateStoreESProducer_cfi

mmusich commented:
> Indeed, I'd remove them for master first, and then backport here.
> I can do the master PR tomorrow (this evening is already too late for me)

https://github.com/cms-sw/cmssw/pull/42656

perrotta commented:
@Dr15Jones https://github.com/cms-sw/cmssw/pull/42656 had been merged in master: could you please backport it here and in the other PRs for 13_1 and 13_0?


In https://github.com/cms-sw/cmssw/pull/42656 a few includes that could be possibly removed were left untouched. Since you, Chris, added them to the original https://github.com/cms-sw/cmssw/pull/42514 PR, maybe you remember why you did it and if it is possible to remove them also. In particular:

In the following configs https://github.com/cms-sw/cmssw/pull/42514 both added the include and inserted the module SiPixelTemplateStoreESProducer in the paths, while subsequent https://github.com/cms-sw/cmssw/pull/42597 removed the module from the paths but not the include from the config:

- FastSimulation/Configuration/python/Reconstruction_BefMix_cff.py
- RecoTracker/FinalTrackSelectors/python/MergeTrackCollections_cff.py

For the following configs, instead, only the include was added by https://github.com/cms-sw/cmssw/pull/42514, but the module therein was never used. I have the impression that it was slipped there by mistake, in particular for the DQM one, but I may be wrong:

- DQMOffline/Trigger/python/SiStrip_OfflineMonitoring_cff.py

- RecoTracker/TransientTrackingRecHit/python/TTRHBuilders_cff.py

Dr15Jones commented:
> maybe you remember why you did it and if it is possible to remove them also. 

So the original change was when the EventSetup module, (i.e. ES module) was assigned to a `cms.Task`. When an ES module is assigned to a cms.Task then all places where that module could be used must also be assigned to a `cms.Task` and that task must be added to the appropriate `Sequence`, `Path` or `EndPath`. So almost all of the configuration changes were to add the appropriate `cms.Task` to a `Sequence` or `Path`. In order to do that, I had to also `import` the cfi.py file, even if that cfi.py might be indirectly included by some other import. When I decided that the use of `cms.Task` was probably a bad idea at this point and removed the `cms.Task`, I failed to also remove the `import`. 

So removing the `import` can be safely done for MOST files but one still must be sure that at least one file in the full job configuration does an `import` so that the ES module is actually added to the job.

Even in the original PR, I tried to have the `import` done in the file which also imports the ES module which uses the products from SiPixelTemplateStoreESProducer. I THINK that was `TTRHBuilderWithTemplate` but that could be wrong and there could be others.

Essentially, I added the `cms.Task` based on what unit tests and RelVal tells failed. I never ran the same test after removing the `cms.Task` and trying to remove the imports.

Dr15Jones commented:
> So, it is probably safe to stay with just the removals listed in https://github.com/cms-sw/cmssw/pull/42656. And at most we'll remain with some extra include here and there. Do you agree?

That's fine with me.

cmsbuild commented:
Pull request #42644 was updated. @Martin-Grunewald, @tjavaid, @perrotta, @civanch, @cmsbuild, @missirol, @mandrenguyen, @pmandrik, @consuegs, @emanueleusai, @mdhildreth, @syuvivida, @sbein, @mmusich, @micsucmed, @francescobrivio, @ssekmen, @nothingface0, @clacaputo, @saumyaphor4252, @rvenditti, @tvami can you please check and sign again.

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-67adef/34528/summary.html
**COMMIT**: 7e323893acfdd01c8d3a15f20404a8028670dfa3
**CMSSW**: CMSSW_13_2_X_2023-08-29-1100/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/42644/34528/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially removed 11 lines from the logs
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3197913
 - DQMHistoTests: Total failures: 3
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3197888
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 47 files compared)
 - Checked 207 log files, 159 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_2_X_2023-08-29-1100+67adef/58644/triggerResults): no differences found

mandrenguyen commented:
+1

mmusich commented:
+hlt

perrotta commented:
+alca

perrotta commented:
+db

emanueleusai commented:
+1

civanch commented:
+1

rappoccio commented:
+1
